pkg/mockclient/mockclient.go

```diff
@@ -0,0 +1,246 @@
+package mockclient
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awssession "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	api "github.com/gardener/machine-controller-manager-provider-aws/pkg/aws/apis"
+)
+
+// MockDriverSPIImpl is the mock implementation of DriverSPI interface that makes dummy calls
+type MockDriverSPIImpl struct {
+	FakeInstances []ec2.Instance
+}
+
+// NewSession starts a new AWS session
+func (ms *MockDriverSPIImpl) NewSession(Secrets api.Secrets, region string) (*awssession.Session, error) {
+	if region == "fail-at-region" {
+		return nil, fmt.Errorf("Region doesn't exist while trying to create session")
+	}
+	return &awssession.Session{}, nil
+}
+
+// NewEC2API Returns a EC2API object
+func (ms *MockDriverSPIImpl) NewEC2API(session *session.Session) ec2iface.EC2API {
+	return &MockEC2Client{
+		FakeInstances: &ms.FakeInstances,
+	}
+}
+
+// MockEC2Client is the mock implementation of an EC2Client
+type MockEC2Client struct {
+	ec2iface.EC2API
+	FakeInstances *[]ec2.Instance
+}
+
+// DescribeImages implements a mock describe image method
+func (ms *MockEC2Client) DescribeImages(input *ec2.DescribeImagesInput) (*ec2.DescribeImagesOutput, error) {
+
+	if *input.ImageIds[0] == "fail-query-at-DescribeImages" {
+		return nil, fmt.Errorf("Couldn't find image with given ID")
+	}
+
+	rootDeviceName := "test-root-disk"
+
+	return &ec2.DescribeImagesOutput{
+		Images: []*ec2.Image{
+			&ec2.Image{
+				RootDeviceName: &rootDeviceName,
+			},
+		},
+	}, nil
+}
+
+// RunInstances implements a mock run instance method
+// The name of the newly created instances depends on the number of instances in cache starts from 0
+func (ms *MockEC2Client) RunInstances(input *ec2.RunInstancesInput) (*ec2.Reservation, error) {
+
+	if *input.ImageId == "fail-query-at-RunInstances" {
+		return nil, fmt.Errorf("Couldn't run instance with given ID")
+	}
+
+	instanceID := fmt.Sprintf("i-0123456789-%d", len(*ms.FakeInstances))
+	privateDNSName := fmt.Sprintf("ip-%d", len(*ms.FakeInstances))
+
+	newInstance := ec2.Instance{
+		InstanceId:     &instanceID,
+		PrivateDnsName: &privateDNSName,
+		State: &ec2.InstanceState{
+			Code: aws.Int64(16),
+			Name: aws.String("running"),
+		},
+		Tags: input.TagSpecifications[0].Tags,
```

Comment:
DeepCopy required?

Reply:
Does this look alright? - https://github.com/gardener/machine-controller-manager-provider-aws/pull/2/commits/846bb51cdf2a10d7b668a130a15afd964f04eb4c